static/js/chess.js

```diff
@@ -42,10 +53,97 @@ document.addEventListener('DOMContentLoaded', function() {
     function startNewGame() {
         gameState = initializeBoard();
         currentPlayer = 'white';
+        isGameOver = false;
         status.textContent = "White's Turn";
         updateBoard();
     }
 
+    function findKingPosition(color) {
+        const kingSymbol = color === 'white' ? '♔' : '♚';
+        for (const [position, piece] of gameState.pieces) {
+            if (piece.piece === kingSymbol) {
+                return position;
+            }
+        }
+        return null;
+    }
+
+    function isKingInCheck(color) {
+        const kingPos = findKingPosition(color);
+        if (!kingPos) return false;
+
+        const oppositeColor = color === 'white' ? 'black' : 'white';
+        for (const [position, piece] of gameState.pieces) {
+            if (piece.color === oppositeColor && isValidMove(position, kingPos, true)) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    function isCheckmate(color) {
+        if (!isKingInCheck(color)) return false;
+        return getAllValidMoves(color).length === 0;
+    }
+
+    function isStalemate(color) {
+        if (isKingInCheck(color)) return false;
+        return getAllValidMoves(color).length === 0;
+    }
+
+    function getAllValidMoves(color) {
+        const moves = [];
+        const squares = board.getElementsByClassName('chess-square');
+        
+        gameState.pieces.forEach((piece, from) => {
+            if (piece.color === color) {
+                Array.from(squares).forEach(square => {
+                    const to = square.dataset.position;
+                    if (isValidMove(from, to, false)) {
+                        // Test if move would leave king in check
+                        const tempState = simulateMove(from, to);
+                        if (!wouldBeInCheck(color, tempState)) {
+                            moves.push({ from, to });
+                        }
+                    }
+                });
+            }
+        });
+        return moves;
+    }
```

Comment:
This is displaying too many moves as being valid. It should only show the actual moves a piece can make. Note - it's actually correctly stopping the player from making the moves, but the green dots show up in places with invalid moves